apps/builder/app/builder/features/ai/apply-operations.ts

```diff
@@ -0,0 +1,143 @@
+import { serverSyncStore } from "~/shared/sync";
+import { generateDataFromEmbedTemplate } from "@webstudio-is/react-sdk";
+import { operations } from "@webstudio-is/ai";
+import { isBaseBreakpoint } from "~/shared/breakpoints";
+import {
+  deleteInstance as _deleteInstance,
+  insertTemplateData,
+} from "~/shared/instance-utils";
+import {
+  breakpointsStore,
+  instancesStore,
+  registeredComponentMetasStore,
+  selectedInstanceStore,
+  styleSourceSelectionsStore,
+  styleSourcesStore,
+  stylesStore,
+} from "~/shared/nano-states";
+import type { DroppableTarget } from "~/shared/tree-utils";
+import { getStyleDeclKey, type StyleSource } from "@webstudio-is/sdk";
+import { nanoid } from "nanoid";
+
+export const applyOperations = (operations: operations.Response) => {
+  for (const operation of operations) {
+    switch (operation.operation) {
+      case "insertTemplate":
+        insertTemplate(operation);
+        break;
+      case "deleteInstance":
+        deleteInstance(operation);
+        break;
+      case "applyStyles":
+        applyStyles(operation);
+        break;
+      default:
+        if (process.env.NODE_ENV === "development") {
+          // eslint-disable-next-line no-console
+          console.warn(`Not supported operation: ${operation}`);
+        }
+    }
+  }
+};
+
+const insertTemplate = (
+  operation: operations.generateInsertTemplate.wsOperation
+) => {
+  const breakpoints = breakpointsStore.get();
+  const breakpointValues = Array.from(breakpoints.values());
+  const baseBreakpoint = breakpointValues.find(isBaseBreakpoint);
+  if (baseBreakpoint === undefined) {
+    return false;
+  }
+  const metas = registeredComponentMetasStore.get();
+  const templateData = generateDataFromEmbedTemplate(
+    operation.template,
+    metas,
+    baseBreakpoint.id
+  );
+
+  // Occasionally the LLM picks a component name as the insertion point.
+  // Instead of throwing the otherwise correct operation we try to fix this here.
+  if (metas.has(operation.addTo)) {
+    const selectedInstance = selectedInstanceStore.get();
+    if (selectedInstance) {
+      operation.addTo = selectedInstance.id;
+    }
+  }
+
+  const rootInstanceIds = templateData.children
+    .filter((child) => child.type === "id")
+    .map((child) => child.value);
+
+  const dropTarget: DroppableTarget = {
+    parentSelector: [operation.addTo],
+    position: operation.addAtIndex + 1,
+  };
+
+  insertTemplateData(templateData, dropTarget);
+  return rootInstanceIds;
+};
+
+const deleteInstance = (operation: operations.deleteInstance.wsOperation) => {
```

Comment:
would be better to have a more precise name than to have the same name but different interfaces and prefixing them, e.g. this seems to delete instance by operation, could be deleteInstanceByOp

Reply:
yes, but they read less nicely, we don't do prefixes generally, instead we use semantic naming to denote the difference